conbench/entities/history.py

```diff
@@ -255,58 +254,63 @@ def get_history_for_cchr(
     return samples
 
 
-def set_z_scores(benchmark_results: List[BenchmarkResult]):
+def set_z_scores(benchmark_results: List[BenchmarkResult], baseline_commit: Commit):
     """Set the "z_score" attribute on each given BenchmarkResult, comparing it to the
-    distribution of BenchmarkResults that share the same case/context/hardware/repo,
-    ending with the most recent BenchmarkResult in the given BenchmarkResult's commit
-    ancestry on the default branch. (We choose a commit from the default branch as the
-    end commit, because we probably don't want "bad" commits early on in a PR to set
-    artificially low z-scores later in the PR.)
+    distribution of BenchmarkResults that share the same case/context/hardware/repo, in
+    the git ancestry of the baseline_commit (inclusive).
+
+    The given benchmark_results must all have the same run_id.
 
     This function does not affect the database whatsoever. It only populates an
     attribute on the given python objects. This is typically called right before
     returning a jsonify-ed response from the API, so the "z_score" attribute shouldn't
     already be set, but if it is already set, it will be silently overwritten.
 
-    Should not raise anything. If we can't find a z-score for some reason, the z_score
-    attribute will be None on that BenchmarkResult.
+    Should not raise anything, except a ValueError if there are mixed run_ids provided.
+    If we can't find a z-score for some reason, the z_score attribute will be None on
+    that BenchmarkResult.
     """
-    # For most invocations of this function, there are very few unique run_ids among the
-    # benchmark_results. Sort them by run_id and run aoptimized query for each group.
-    sorted_by_run_id: DefaultDict[str, List[BenchmarkResult]] = defaultdict(list)
+    run_ids = set(result.run_id for result in benchmark_results)
+    if len(run_ids) != 1:
+        raise ValueError(f"Encountered mixed run_ids in set_z_scores(): {run_ids}")
+    run_id = run_ids.pop()
+
+    if len(benchmark_results) == 1:
+        # performance optimization
+        case_id = benchmark_results[0].case_id
+        context_id = benchmark_results[0].context_id
+    else:
+        case_id = None
+        context_id = None
+
+    distribution_stats = _query_distribution_stats_by_run_id(
```

Comment:
Future: let's think about the name `_query_distribution_stats_by_run_id()` and/or add a code comment. It takes too long for me to figure out what this is doing :).

What `_query_distribution_stats_by_run_id()` actually does. But when reading the code plain as-is, I am confused --- what would the distribution for a Run be, considering that a Run largely is "just" a group (of potentially super unrelated) BenchmarkResults?

Does this operate on all BenchmarkResults associated with that Run?

Reply:
Ah shoot, I was going to change the name of this function in this PR, and forgot. Thanks for calling this out! 

Run ID is much less important to this function now than it used to be. It used to absolutely determine the distribution. Now the distribution is mostly determined by the baseline commit (and results are filtered to the run's hardware).